### PR TITLE
feat: Add install script for Vector

### PR DESCRIPTION
**Pull request by spencergilbert**
Opened as an alternative to https://github.com/vectordotdev/vector/pull/18777

This struck me as a better location to maintain the script as AP is likely to be responsible for it as it is for the Agent and OPW scripts. Additionally since the scripts are so ~~copied~~ related co-locating them seems sensible.

[Original PR](https://github.com/vectordotdev/vector/pull/18777)

### CODE REVIEW

File: install_script_vector0.sh [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+else
+    sudo_cmd='sudo'
+fi
+
+# Install the necessary package sources
+if [ "$OS" == "RedHat" ]; then
+    echo -e "\033[34m\n* Installing YUM sources for Vector\n\033[0m"
+
+    if "$cloudsmith_migrate"; then
+        rm "/etc/yum.repos.d/timber-vector.repo"
```
Comment (chouquette) [inline install_script_vector0.sh:128]: maybe worth adding a `-f` in case the removal is attempted a 2nd time

File: install_script_vector0.sh [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    for key_path in "${RPM_GPG_KEYS[@]}"; do
+      gpgkeys="${gpgkeys:+"${gpgkeys}${separator}"}https://${keys_url}/${key_path}"
+    done
+
+    $sudo_cmd sh -c "echo -e '[vector]\nname = Vector\nbaseurl = https://${yum_url}/${yum_version_path}/\$basearch/\nenabled=1\ngpgcheck=1\nrepo_gpgcheck=${rpm_repo_gpgcheck}\npriority=1\ngpgkey=${gpgkeys}' > /etc/yum.repos.d/vector.repo"
+
+    $sudo_cmd yum -y clean metadata
+elif [ "$OS" == "Debian" ]; then
+    if "$cloudsmith_migrate"; then
+        rm "/etc/apt/sources.list.d/timber-vector.list"
```
Comment (chouquette) [inline install_script_vector0.sh:152]: ditto for `-f`

File: install_script_vector0.sh [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    do
+        printf "\033[34m\n* Installing apt-transport-https, curl and gnupg\n\033[0m\n"
+        $sudo_cmd apt-get update || printf "\033[31m'apt-get update' failed, the script will not install the latest version of apt-transport-https.\033[0m\n"
+        # installing curl might trigger install of additional version of libssl; this will fail the installation process,
+        # see https://unix.stackexchange.com/q/146283 for reference - we use DEBIAN_FRONTEND=noninteractive to fix that
+        apt_exit_code=0
+
+        $sudo_cmd sh -c "DEBIAN_FRONTEND=noninteractive apt-get install -y apt-transport-https curl gnupg 2>$VEC_APT_INSTALL_ERROR_MSG || apt_exit_code=$?"
+
+        if grep "Could not get lock" $VEC_APT_INSTALL_ERROR_MSG; then
```
Comment (chouquette) [inline install_script_vector0.sh:170]: I realize this is not at all related to your PR, but how does this check behaves with i18n?

File: install_script_vector0.sh [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    MAX_RETRY_NB=10
+    for i in $(seq 1 $MAX_RETRY_NB)
+    do
+        printf "\033[34m\n* Installing apt-transport-https, curl and gnupg\n\033[0m\n"
+        $sudo_cmd apt-get update || printf "\033[31m'apt-get update' failed, the script will not install the latest version of apt-transport-https.\033[0m\n"
+        # installing curl might trigger install of additional version of libssl; this will fail the installation process,
+        # see https://unix.stackexchange.com/q/146283 for reference - we use DEBIAN_FRONTEND=noninteractive to fix that
+        apt_exit_code=0
+
+        $sudo_cmd sh -c "DEBIAN_FRONTEND=noninteractive apt-get install -y apt-transport-https curl gnupg 2>$VEC_APT_INSTALL_ERROR_MSG || apt_exit_code=$?"
```
Comment (chouquette) [inline install_script_vector0.sh:168]: Couldn't we `export DEBIAN_FRONTEND=noninteractive` and invoke with a regular `$sudo_cmd apt...` ?

File: install_script_vector0.sh [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        elif [ $apt_exit_code -ne 0 ]; then
+            cat $VEC_APT_INSTALL_ERROR_MSG
+            exit $apt_exit_code
+        else
+            break
+        fi
+    done
+
+    printf "\033[34m\n* Installing APT package sources for Vector\n\033[0m\n"
+    $sudo_cmd sh -c "echo 'deb [signed-by=${apt_usr_share_keyring}] https://${apt_url}/ ${apt_repo_version}' > /etc/apt/sources.list.d/vector.list"
```
Comment (chouquette) [inline install_script_vector0.sh:183]: Wouldn't `"echo 'deb [signed-by=${apt_usr_share_keyring}] https://${apt_url}/ ${apt_repo_version}' | $sudo_cmd tee 
/etc/apt/sources.list.d/vector.list` be more idiomatic?

File: install_script_vector0.sh (hunk 3 above)
Comment (spencergilbert) [reply install_script_vector0.sh:170]: 😓 probably poorly, but it doesn't appear to have come up for the "upstream" scripts yet? Might be worth opening an issue to track though.

File: install_script_vector0.sh (hunk 5 above)
Comment (spencergilbert) [reply install_script_vector0.sh:183]: I'm ambivalent, happy to test and upstream the change though.

File: install_script_vector0.sh (hunk 4 above)
Comment (spencergilbert) [reply install_script_vector0.sh:168]: I don't see why not - I can test it and upstream it to the template.

File: install_script_vector0.sh (hunk 1 above)
Comment (spencergilbert) [reply install_script_vector0.sh]: Addressed in c082780

File: install_script_vector0.sh (hunk 2 above)
Comment (spencergilbert) [reply install_script_vector0.sh]: Addressed in c082780

File: install_script_vector0.sh (hunk 4 above)
Comment (spencergilbert) [reply install_script_vector0.sh]: Addressed in c082780

File: install_script_vector0.sh (hunk 5 above)
Comment (spencergilbert) [reply install_script_vector0.sh]: Addressed in c082780

File: install_script_vector0.sh (hunk 3 above)
Comment (chouquette) [reply install_script_vector0.sh:171]: Oh yeah this is definitely a question for the upstream version, but I didn't thought of it at the time :grimacing:

**Review by chouquette — APPROVED** [review]

File: install_script_vector0.sh (hunk 3 above)
Comment (chouquette) [reply install_script_vector0.sh:171]: https://datadoghq.atlassian.net/browse/APL-2224 created for that purpose

**Pull request merged**